src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php

```diff
@@ -83,10 +86,47 @@ public function accept(Definition $definition): bool
         return 80000 <= \PHP_VERSION_ID && $definition->isAutoconfigured() && !$definition->hasTag($this->ignoreAttributesTag);
     }
 
-    public function processClass(ContainerBuilder $container, \ReflectionClass $class)
+    public function registerAutoconfigureAttributes(ContainerBuilder $container, \ReflectionClass $class): void
     {
         foreach ($class->getAttributes(Autoconfigure::class, \ReflectionAttribute::IS_INSTANCEOF) as $attribute) {
             ($this->registerForAutoconfiguration)($container, $class, $attribute);
         }
     }
+
+    public function registerServiceAttributes(ContainerBuilder $container, \ReflectionClass $class, string $id = null): string
+    {
+        $typedServices = [];
+
+        foreach ($class->getAttributes(Service::class, \ReflectionAttribute::IS_INSTANCEOF) as $attribute) {
+            /** @var Service $service */
+            $service = $attribute->newInstance();
+
+            if ($service->type) {
+                if ($service->type !== $class->name && !$class->isSubclassOf($service->type)) {
+                    throw new InvalidArgumentException(sprintf('Type "%s" in #[Service] attribute should be a supertype of class "%s".', $service->type, $class->name));
```

Comment:
The error message does not tell me that `$service->type === $class->name` would also be acceptable.

Reply:
Yeah, I think the current message is good enough. At least I don't have a better idea that wouldn't end up being too long for such a message.